ssm_cache/cache.py

```diff
@@ -5,16 +5,101 @@
 from functools import wraps
 import six
 
-import boto3
-
 class InvalidParameterError(Exception):
     """ Raised when something's wrong with the provided param name """
 
+class SSMParameterStore(object): # pylint: disable=too-few-public-methods
+    """Some hooks are provided to override behavior: see inject_session_factory and inject_client_factory
+    for details."""
+
+    @classmethod
+    def _default_session_factory(cls):
+        """Default session factory that creates a boto3 session."""
+        import boto3
+        return boto3.session.Session()
+
+    @classmethod
+    def _default_client_factory(cls, session, name):
+        """Default client factory that creates a client from the provided session."""
+        return session.client(name)
+
+    _SESSION_FACTORY = _default_session_factory
+    _CLIENT_FACTORY = _default_client_factory
+
+    _SESSION = None
+    _CLIENT = None
+
+    @classmethod
+    def inject_session_factory(cls, session_factory):
+        """Takes a function that takes no parameters and returns an object that acts
+        like a boto3 session. If this value is not None, it is used instead of creating
+        a regular boto3 session. This could be made to use placebo
+        (https://github.com/garnaat/placebo) or another library for testing.
+
+        **The provided factory method must be a static method.***
+
+        Returns the previous session factory."""
+        old_factory = cls._SESSION_FACTORY
+        cls._SESSION_FACTORY = staticmethod(session_factory)
+        return old_factory
+
+    @classmethod
+    def inject_client_factory(cls, client_factory):
+        """Takes a function that takes no parameters and returns an object that acts
+        like a boto3 SSM client. If this value is not None, it is used instead of creating
+        a regular boto3 client.
+
+        **The provided factory method must be a static method.***
+
+        Returns the previous client factory."""
+        old_factory = cls._CLIENT_FACTORY
+        cls._CLIENT_FACTORY = staticmethod(client_factory)
+        return old_factory
+
+    @classmethod
+    def _session(cls):
+        """Use the defined session factory to create an object that acts like a boto3 session.
+        Defaults to boto3.session.Session(); set SESSION_FACTORY to inject a different session
+        factory."""
+        if cls._SESSION is None:
+            if cls._SESSION_FACTORY:
+                cls._SESSION = cls._SESSION_FACTORY()
+            else:
+                cls._SESSION = cls._default_session_factory()
+        return cls._SESSION
+
+    @classmethod
+    def _client(cls):
+        """Use the defined client factory to create an object that acts like a boto3 client.
+        Defaults to _session().client("ssm"); set CLIENT_FACTORY to inject a different client
+        factory."""
+        if cls._CLIENT is None:
+            if cls._CLIENT_FACTORY:
+                client = cls._CLIENT_FACTORY(cls._session(), "ssm")
```

Comment:
Here I would pass named arguments:

```python
client = cls._CLIENT_FACTORY(
    session=cls._session(),
    name="ssm",
)
```

This would allow users to provide complex factories, with additional default arguments and/or order-invariant arguments. Something like:

```python

class MyComplexClientFactory(object):
    def __init__(self, name, session, my_optional_param=5):
        # whatever they need to do
```

Reply:
I'm not certain this will work given the use of `staticmethod` to get around the python magic but if you say it works...

Reply:
It will be worth a test :) I'll see what we can do about that `staticmethod` trick, I'll be happy to remove it if possible.